### PR TITLE
13.0 opw 2322735 fix agr

### PR DESCRIPTION
**Pull request by agr-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by agr-odoo** [comment]
@nim-odoo 
Does this makes sense in your opinion?
If this is a too big change to merge I'll stop losing time on it...
it still a draft (with bank payments it does not work for some reason)
Note that "[FIX] point_of_sale: partner_id on split move lines" is already merged in master (https://github.com/odoo/odoo/pull/56142 pim gave his blessing, also for v13 but at the time was not really needed)

File: addons/point_of_sale/models/pos_session.py [hunk 1]
```diff
@@ -380,7 +380,10 @@ def _accumulate_amounts(self, data):
 
             if order.is_invoiced:
                 # Combine invoice receivable lines
-                key = order.partner_id.property_account_receivable_id.id
+                if order.payment_ids[:1].payment_method_id.split_transactions:
+                    key = (order.partner_id.property_account_receivable_id.id, payment)
+                else:
+                    key = order.partner_id.property_account_receivable_id.id
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_session.py:386]: Keep the key type consistent:
```suggestion
                    key = (order.partner_id.property_account_receivable_id.id, False)
```

File: addons/point_of_sale/models/pos_session.py [hunk 2]
```diff
@@ -542,12 +545,19 @@ def _create_invoice_receivable_lines(self, data):
...
         invoice_receivable_lines = {}
-        for receivable_account_id, amounts in invoice_receivables.items():
-            invoice_receivable_vals[receivable_account_id].append(self._get_invoice_receivable_vals(receivable_account_id, amounts['amount'], amounts['amount_converted']))
+        for key, amounts in invoice_receivables.items():
+            try:
+                receivable_account_id, payment = key
+                partner_id = payment.partner_id.id
+            except TypeError:
+                receivable_account_id = key
+                partner_id = False
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_session.py:554]: With consistent type you don't need the Try/except:
```suggestion
            receivable_account_id, payment = key
            partner_id = payment.partner_id.id if payment else False
```

**Review by nim-odoo — COMMENTED** [review]
I think this is safer to do it at this level rather than in the Partner Ledger. Moreover, the payment of an invoice is consistent between the POS and the backend. This is a slight change of behavior, but I think the current one is confusing: from what I tested, it seems more logical with your fix.

I think the impact should be low: it only affects invoiced orders with payments in 'split' configuration.

robodoo rebase-ff

**Comment by agr-odoo** [comment]
I've added the relevant changes, but there is still some issue:
- when using a non-cash payment type the partner ledger is not balanced

Here it is the journal entry of a POS session in which:
- CustomerC pay 12.50 via cash
- CustomerB pay 70.00 via bank
after_fix_je

Partner ledger of CustomerC
after_fix_custC

Partner ledger of CustomerB
after_fix_custB

CustomerB has unbalanced entries because in the POS journal entries it seems that the line where the bank account of the customer is debited and the receivable credited are missing and I don't know if we could add it.
So as an alternative there is also https://github.com/odoo/odoo/pull/58937 in which the partner_id assignment has been made more convenient

Here it is the journal entry of the same POS session, with same customer settings
after_fix_je_1

Partner ledger of CustomerC
after_fix_custC_1

Partner ledger of CustomerB
after_fix_custB_1

**Comment by nim-odoo** [comment]
Hum, I don't get why the cash and the bank entries are handled differently in a first place. If I'm not wrong, cash and bank are pretty much equivalent from an accounting perspective. I'd be tempted to say that the correct entries are the cash entries, so you should apply `_get_split_receivable_vals` to the bank entries. Maybe we can ask TSB his opinion before going further.

**Comment by agr-odoo** [comment]
Hello @pimodoo 
May I ask your opinion for this not-so-small change?
Summary: The accounting issue with the partner ledger would be fixed

Below you can see an example of session in which
- CustomerC pay 70.00 via cash
- CustomerB pay 12.50 via bank
For cash payment everything is fine, for bank I guess the user was required to do a last step of manual reconciliation because a bank entry is not automatically created like in cash use case

after_fix_je
after_fix_custC

**Comment by agr-odoo** [comment]
@robodoo up to 14.0

**Comment by agr-odoo** [comment]
@fw-bot up to 14.0

**Comment by pimodoo** [comment]
@agr-odoo Are you sure it is still needed? The partner ledger has been fixed in https://github.com/odoo/enterprise/pull/13387

The difference between cash and bank is that the bank entry is made when the users are importing their bank statements. Which is the only correct way to encode bank entries (or doing manual paiements)

FYI @nim-odoo

**Comment by pimodoo** [comment]
So we could probably keep the commit that set a partner on the line (if calling _find_accounting_partner with 'false' value works as most of the time there are no partners)

But I wouldn't keep the other commit

**Comment by pimodoo** [comment]
I also don't understand why forwardporting only to 14.0?

**Comment by agr-odoo** [comment]
Are you sure it is still needed? The partner ledger has been fixed in [odoo/enterprise#13387](https://github.com/odoo/enterprise/pull/13387)
> 
Ah crap, thanks for the info, I'll check in a moment

> The difference between cash and bank is that the bank entry is made when the users are importing their bank statements. Which is the only correct way to encode bank entries (or doing manual paiements)

Ok, thanks, nice to know

> So we could probably keep the commit that set a partner on the line (if calling _find_accounting_partner with 'false' value works as most of the time there are no partners)

> I also don't understand why forwardporting only to 14.0?

 "[FIX] point_of_sale: partner_id on split move lines" was already merged in master, so now should be in 14.0 #56142

**Comment by pimodoo** [comment]
ok, nice :)

**Comment by agr-odoo** [comment]
Are you sure it is still needed? The partner ledger has been fixed in [odoo/enterprise#13387](https://github.com/odoo/enterprise/pull/13387)
> 

I've just checked in a runbot v13 (https://4485076-13-0.runbot52.odoo.com/)
and I see the same issue

**Partner ledger when not issuing invoice**
runbot-13-noinv
**Partner ledger with invoice**
runbot-13

The fix you mentioned before add the 'unkown' section to collect all entries without partner but the user would need to do a lot of bookkeeping
runbot-13-unkown-e

**Comment by pimodoo** [comment]
What you are showing, are the filters of the account move lines not the partner ledger report, you have to go in the menu "reporting" -> "Partner Ladger" (not "accounting" -> "partner ledger". Maybe that's what you are looking for.

https://tinyurl.com/yyv6f2e3

**Comment by agr-odoo** [comment]
> What you are showing, are the filters of the account move lines not the partner ledger report, you have to go in the menu "reporting" -> "Partner Ladger" (not "accounting" -> "partner ledger". Maybe that's what you are looking for.
> 

Nice, yes, from that view the partner entries looks ok
The incomplete ledger accessed from the contact form smart button could be now considered a minor issue? 
cc @nim-odoo

**Pull request closed without merging**